packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts

```diff
@@ -47,12 +79,37 @@ export class EvmModule {
 
   // evm_mine
 
-  private _mineParams(params: any[]): [] {
-    return validateParams(params);
+  private _mineParams(params: any[]): [number] {
+    if (params.length === 0) {
+      params.push(0);
+    }
+    return validateParams(params, t.number);
   }
 
-  private async _mineAction(): Promise<string> {
-    await this._node.mineEmptyBlock();
+  private async _advanceTimeOffsetAccordingToTimestamp(timestamp: BN) {
+    const latestBlock = await this._node.getLatestBlock();
+    const increment = new BN(timestamp).sub(
+      new BN(latestBlock.header.timestamp)
+    );
+    if (increment.lte(new BN(0))) {
+      throw new InvalidInputError(
+        `Timestamp ${timestamp} is lower than previous block's timestamp`
+      );
+    }
+    await this._node.increaseTime(increment);
+  }
+
+  private async _mineAction(timestamp: number): Promise<string> {
+    if (timestamp !== 0) {
+      await this._advanceTimeOffsetAccordingToTimestamp(new BN(timestamp));
+    } else {
+      const nextBlockTimestamp = await this._node.getNextBlockTimestamp();
+      if (!nextBlockTimestamp.eq(new BN(0))) {
```

Comment:
I don't understand this `if`. Why is this necessary? To ensure that the next block is mined with `nextBlockTimestamp`?

Reply:
It checks to ensure `nextBlockTimestamp` is set. If it is not (value is 0), it means the next block time Is set and it set the next block timestamp to `nextBlockTimestamp`.